packages/core/src/services/loopDetectionService.ts

```diff
@@ -118,12 +156,85 @@ export class LoopDetectionService {
     return false;
   }
 
+  private async checkForLoopWithLLM(signal: AbortSignal) {
+    const recentHistory = this.config
+      .getGeminiClient()
+      .getHistory()
+      .slice(-LLM_LOOP_CHECK_HISTORY_COUNT);
+
+    const prompt = `You are a sophisticated AI diagnostic agent specializing in identifying when a conversational AI is stuck in an unproductive state. Your task is to analyze the provided conversation history and determine if the assistant has ceased to make meaningful progress.
+
+An unproductive state is characterized by one or more of the following patterns over the last 5 or more assistant turns:
+
+Repetitive Actions: The assistant repeats the same tool calls or conversational responses a decent number of times. This includes simple loops (e.g., tool_A, tool_A, tool_A) and alternating patterns (e.g., tool_A, tool_B, tool_A, tool_B, ...).
+
+Cognitive Loop: The assistant seems unable to determine the next logical step. It might express confusion, repeatedly ask the same questions, or generate responses that don't logically follow from the previous turns, indicating it's stuck and not advancing the task.
+
+Crucially, differentiate between a true unproductive state and legitimate, incremental progress.
+For example, a series of 'replace' or 'write_file' tool calls that make small, distinct changes to the same file (like adding docstrings to functions one by one) is considered forward progress and is NOT a loop. A loop would be repeatedly replacing the same text with the same content, or cycling between a small set of files with no net change.
+
+Please analyze the conversation history to determine the possibility that the conversation is stuck in a repetitive, non-productive state.
+
+**Conversation History:**:
```

Comment:
You shouldn't need to provide conversation history in this way. You can pass this to the `generateJson` method. See the nextSpeakerChecker as context if you need